sdx_controller/controllers/connection_controller.py

```diff
@@ -57,12 +58,21 @@ def place_connection(body):
     :rtype: Connection
     """
     logger.info(f"Placing connection: {body}")
-    if connexion.request.is_json:
-        body = connexion.request.get_json()
-        logger.info(f"Gathered connexion JSON: {body}")
+    if not connexion.request.is_json:
+        return "Request body must be JSON", 400
+
+    body = connexion.request.get_json()
+    logger.info(f"Gathered connexion JSON: {body}")
 
     logger.info("Placing connection. Saving to database.")
-    db_instance.add_key_value_pair_to_db("connection_data", json.dumps(body))
+
+    if "id" in body:
+        connection_id = body["id"]
```

Comment:
Trusting that clients would provide a unique ID for each request is problematic. What will happen if I send a series of identical connection requests with the same ID, by mistaken configuration, resent messages (because controller took a long time to respond), or plain bad intent?

Reply:
If SDX generates a uuid for each connection, then it will need to respond with the connection id to the client. Because the client (MEICAN in this case) will use that id to query or delete the connection in the future. This can cause issue: What if the client didn't receive the response that contains the id?

We might be able to add a checking part to verify if the user provided a uuid, else return a 400 with "connection must be uuid" prompt.

Reply:
Can't we do a DB lookup to find duplicates?

If I understand correctly, Swagger UI will reject invalid requests, true, but that validation happens only in the browser using Javascript. Does the `POST /connection` endpoint proper reject invalid inputs though? I doubt we do that validation. We should test this.

Reply:
> Can't we do a DB lookup to find duplicates?

We can find duplicates from DB lookup. The current issue is that, since we don't have an update_connection endpoint, for now, the only way to update connection is via place_connection endpoint with same id (not implemented yet). So I meant, we can look for duplicate, but cannot reject request with duplicate id yet. 

If we decide to add a update_connection, we can reject duplicate id request.
> 
> If I understand correctly, Swagger UI will reject invalid requests, true, but that validation happens only in the browser using Javascript. Does the `POST /connection` endpoint proper reject invalid inputs though? I doubt we do that validation. We should test this.

Ya, I tried, it rejects requests without a id.
```
{
  "detail": "'id' is a required property",
  "status": 400,
  "title": "Bad Request",
  "type": "about:blank"
}
```

Reply:
the /connection endpoints supports 'put' method, like others, where put /connection/{connection_id} should be interpreted as "update".

Reply:
Can't we do a DB lookup in `POST /connection` handler? That way if I send a sequence of identical connection requests, the first one will get processed, and the rest will be rejected as duplicates.  This could even be a test case.

Reply:
REST POST is supposed to be idempotent? ie, only the first one of multiple requests would be processed? 
Yes, it's worth a test cast

Reply:
> the /connection endpoints supports 'put' method, like others, where put /connection/{connection_id} should be interpreted as "update".

Currently there's only POST for connection; GET and DELETE for connection/{connection_id}, code [here](https://github.com/atlanticwave-sdx/sdx-controller/blob/233afcb2a9c2c436136276c4d987fb2bfc17e74b/sdx_controller/swagger/swagger.yaml#L137-L216). We should probably add an PUT endpoint to update an existing connection.

Reply:
> REST POST is supposed to be idempotent? ie, only the first one of multiple requests would be processed? Yes, it's worth a test cast

POST is not naturally idempotent (https://developer.mozilla.org/en-US/docs/Glossary/Idempotent). But since we will add the update connection method, we can make the POST connection API idempotent. Each time the client generates an uuid as connection_id, then server can use that connection_id to identify if this is an duplicate request or not.

Reply:
> Can't we do a DB lookup in `POST /connection` handler? That way if I send a sequence of identical connection requests, the first one will get processed, and the rest will be rejected as duplicates. This could even be a test case.

Yes, since we are adding PUT /connection/{connection_id} API, we can use the client's connection_id to identify duplicate requests.

Reply:
Sorry, typo: meant to see "Put is idempotent", because you're discussing " if I send a sequence of identical connection requests"

Reply:
Oh that's perfect. PUT is idempotent. I'll merge this one for now, then open a new PR to add a PUT /connection/{connection_id}, and we should be all set.